src/libtai/caldate_mjd.3
.TH caldate_mjd 3
.SH NAME
caldate_mjd \- manipulate calendar dates
.SH SYNTAX
.B #include <caldate.h>

void \fBcaldate_frommjd\fP(&\fIcd\fR,\fIday\fR,\fIweekday\fR,\fIyearday\fR);
.br
long \fBcaldate_mjd\fP(&\fIcd\fR);
.br
void \fBcaldate_normalize\fP(&\fIcd\fR);

void \fBcaldate_easter\fP(&\fIcd\fR);

struct caldate \fIcd\fR;
.br
long \fIday\fR;
.br
int *\fIweekday\fR;
.br
int *\fIyearday\fR;
.SH "BASIC CONVERSIONS"
Every calendar date has a
.I modified Julian day number\fR.
The day number increases by 1 every day.
Day number 0 is 17 November 1858.
Day number 51604 is 1 March 2000.

.B caldate_frommjd
puts into
.I cd
the date corresponding to the modified Julian day number
.IR day .

.B caldate_frommjd
also computes the day of the week (0 through 6)
and the day of the year (0 through 365).
It stores the day of the week in
.B *\fIweekday
if
.I weekday
is nonzero.
It stores the day of the year in
.B *\fIyearday
if
.I yearday
is nonzero.

.B caldate_mjd
returns the modified Julian day number corresponding to the date in
.IR cd .

.B caldate_mjd
accepts days outside the range 1 to 31,
referring to days before the beginning or after the end of the month.
It also accepts months outside the range 1 to 12,
referring to months before the beginning or after the end of the year.

.B caldate_normalize
calls
.B caldate_frommjd
with the result of
.BR caldate_mjd .
This means that it accepts out-of-range months and out-of-range days in
.IR cd ,
and puts a valid calendar date back into
.IR cd .
.SH "OTHER FUNCTIONS"
.B caldate_easter
reads the year from
.I cd
and then changes
.I cd
to the date of Easter in the same year.
.SH LIMITATIONS
The
.B caldate
routines currently support the Gregorian calendar,
which was defined in 1582 and adopted at different times
in different countries.
For earlier dates the
.B caldate
routines work with
``virtual Gregorian,''
defined mathematically by the 400-year Gregorian cycle
for years before 1582.
The Julian calendar is not supported.

The Gregorian calendar will be replaced by a new calendar
within a few thousand years.
The
.B caldate_frommjd
and
.B caldate_mjd
routines will be upgraded accordingly.
The current
.B caldate_frommjd
and
.B caldate_mjd
routines are not useful for dates far in the future.

Day numbers will overflow a 32-bit
.B long
sometime after the year 5000000;
all systems should upgrade to 64-bit
.BR long s
before then.
.B caldate_mjd
does not check for overflow.
.SH "SEE ALSO"
caldate(3)


ChangeLog
			SWI-Prolog ChangeLog
			====================

					--------------------------------------
					NOTE: Bug #<N> refers to bugs reported
					to the SWI-Prolog Bugzilla system at
					http://gollem.science.uva.nl/bugzilla/
					--------------------------------------

===================================
VERSION 5.6.19 (September 20, 2006)
===================================

  * MODIFIED: More efficient DCG translation.  No longer produces 'C'/3
    calls.  'C'/3 has been moved to backcomp.pl

  * FIXED: Bug#205: support GNU-make when not named make.  Marc Espie.

  * FIXED: Bug 204: ** for rationals.  Mike Elston.

  * ADDED: Library heaps.pl.  Adapted by Markus Triska.

  * FIXED: Renamed depreciated --> deprecated.  Mike Elston.

================================
VERSION 5.6.18 (August 24, 2006)
================================

  * FIXED: format_time/3 can produce inconsistent results on first call.
    Keri Harris.

  * FIXED: Updated www_open_url for netscape compatible browsers on Unix.
    Cleanup of comments, etc.  Work-around lacking exit code of firefox
    -remote commands.

  * FIXED: Source-level debugger for a --> { x }, !, y.

  * FIXED: write_canonical(a|b).  Lukas Degener.

  * ADDED: Support for large integers in succ/2, plus/3 and between/3.
    Bart Demoen.

  * FIXED: Domain error on random/1.   Lukas Degener.

  * FIXED: pl-tai.c portability problem for Solaris (asctime_r). 
    Hazita Harun.

  * FIXED: Bug#192: reading non-decimal numbers that cause a 64-bit int
    overflow.  Marius Filip.

  * ENHANCED: Bug#195: library(debug) to compile debugging(x) to fail when -O
    is active.  Holger Kanwischer.

  * ADDED: print_message(silent, toplevel_goal(Goal, Bindings)) to allow
    intercepting toplevel goals just before they are executed.  Some
    cleanup to toplevel.pl and topvars.pl

  * FIXED: possible crash freeing a thread that could not be created
    due to conflicting alias.

  * FIXED: avoid errors in open while getting a stacktrace.  Mike
    Elston. 

  * FIXED: 1 is sin(pi/2) incorrectly succeeded in optimised mode.

  * MODIFIED: Message term of make/0 after completion has changed from
    make(done) to make(done(Files)).  Should not affect user-code.

  * ADDED: Library option, option/2.

  * DOC: Fixed all libraries to have their comments compatible with the
    emerging documentation package.

  * FIXED: stead-fastness of library predicate option/3.

  * PORT: Bug#192: Better implementation of per-thread CPUtime, depending
    on Linux library and kernel versions.  Keri Harris.

  * FIXED: Bug#191: with_output_to and format: lacking initialisation.
    Holger Kanwischer.

  * FIXED: Interaction problem between comments and subterm-positions
    options in read_term/3.  Lukas Degener.

  * ADDED: library prolog_source.pl.  Start to generalise opening, closing
    and dealing with operators to handle Prolog source documents outside the
    compiler.

  * FIXED: read_term/2,3 comments(-Comments) option loosing first
    character after /*.

  * FIXED: Bug#189: Thread id in Linux/PPC (and possibly others).
    Keri Harris.

  * FIXED: Bug#188: missing module_transparent for sformat/2,3.  
    Holger Kanwischer.

  * FIXED: Bug#185: %+ initialization if yday and yday.  Keri Harris.

  * FIXED: Bug#184: date/time test-suite for ~ -> %.  Keri Harris.

==============================
VERSION 5.6.17 (July 27, 2006)
==============================

  * MODIFIED: predicate_property(Pred, number_of_clauses(N)) returns 0
    for dynamic predicates with 0 clauses.

  * FIXED: Added Unicode support to DDE interface.

  * FIXED: Various long/int issues calling PL_unify_term().  Affects 64-bit
    systems only.

  * FIXED: Memory violation getting thread-CPU time in Linux.  With help
    of Jacco van Ossenbruggen.

  * ADDED: Allow reading comments in read_term/3.  Ongoing work.

  * FIXED: Graphical debugger after throw/1.  Alan Baljeu.

  * FIXED: fatal error on read/1 in 0-bytes outside quoted strings/atoms.

  * FIXED: Error tracing multi-threaded code in plcon.exe.  Mike Elston.

  * ADDED: Prolog flag verbose_load to get status of silent(Bool) from
    load_files/2.  Tom Schrijvers.

  * ADDED: First version of new time support.  With input from various
    users from the mainlinglist.

  * ENHANCED: Try to reuse variable names when writing results of toplevel
    queries.

  * ADDED: with_output_to(+Out, :Goal) to allow writing to atoms, strings,
    code-lists, etc.  Also allows for format(+Out, +Fmt, +Args) with the
    same specifications.  Moved sformat/[2,3] to the compatibility library.

  * MODIFIED: functions lsb and msb throw domain error on 0.  Updated
    documentation.  With help from Richard O'Keefe and Bart Demoen.

  * ADDED: Documentation for library(ordsets).

==============================
VERSION 5.6.16 (July 05, 2006)
==============================

  * FIXED: Possibly corrupted stack (invalid program counter) returned
    through the library prolog_stack.

==============================
VERSION 5.6.15 (June 29, 2006)
==============================

  * FIXED: Replace hard-coded message about threads not dying by calling
    print_message/2.  Paulo Moura.

  * FIXED: Clear pending exception on `retry'.

  * MODIFIED: threads/0 no longer joins terminated threads.  A new predicate
    join_threads/0 does this.   Paulo Moura.

  * FIXED: Bug#174: print problem reporting problems with a predicate (/)/2.
    Holger Kanwischer.

  * FIXED: possibly lacking EXIT instruction at the end of a clause, causing
    the system to execute arbitrary code.  Notably affects non-GCC
    platforms, but in theory it can also affect GCC compiled systems.  
    Mike Elston.

  * FIXED: Quick load/sated state format handling of MPZ integers.  Mike
    Elston.

  * FIXED: ERROR: atom_codes/2: Cannot represent due to `character_code'
    error in library(url).

==============================
VERSION 5.6.14 (June 13, 2006)
==============================


src/libtai/caldate_mjd.c
#include "tai.h"
#include "caldate.h"

static unsigned long times365[4] = { 0, 365, 730, 1095 } ;
static unsigned long times36524[4] = { 0, 36524UL, 73048UL, 109572UL } ;
static unsigned long montab[12] =
{ 0, 31, 61, 92, 122, 153, 184, 214, 245, 275, 306, 337 } ;
/* month length after february is (306 * m + 5) / 10 */

long caldate_mjd(struct caldate *cd)
{
  long y;
  long m;
  long d;

  d = cd->day - 678882L;
  m = cd->month - 1;
  y = cd->year;

  d += 146097L * (y / 400);
  y %= 400;

  if (m >= 2) m -= 2; else { m += 10; --y; }

  y += (m / 12);
  m %= 12;
  if (m < 0) { m += 12; --y; }

  d += montab[m];

  d += 146097L * (y / 400);
  y %= 400;
  if (y < 0) { y += 400; d -= 146097L; }

  d += times365[y & 3];
  y >>= 2;

  d += 1461L * (y % 25);
  y /= 25;

  d += times36524[y & 3];

  return d;
}
